src/core/search/search.cc

```diff
@@ -25,65 +27,142 @@ AstExpr ParseQuery(std::string_view query) {
   return driver.Take();
 }
 
-// Add results from matched to current, keep sorted and remove duplicates.
-void UnifyResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
-  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
-  current->clear();
-  current->reserve(matched.size() + tmp->size());
+// Represents an either owned or non-owned result set that can be accessed transparently.
+struct IndexResult {
+  using DocVec = vector<DocId>;
+
+  IndexResult() : value{DocVec{}} {};
+
+  IndexResult(const DocVec* dv) : value{dv} {
+    if (dv == nullptr)
+      value = DocVec{};
+  }
+
+  IndexResult(DocVec&& dv) : value{move(dv)} {
+  }
+
+  // Transparent const access to underlying value
+  const DocVec* operator->() const {
+    return holds_alternative<DocVec>(value) ? &get<DocVec>(value) : get<const DocVec*>(value);
+  }
+
+  // Move out of owned or copy borrowed
+  DocVec Take() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return *get<const DocVec*>(value);
+  }
+
+  // Move out of owned or return empty
+  DocVec TakeOwnedOrEmpty() {
+    if (holds_alternative<DocVec>(value))
+      return move(get<DocVec>(value));
+    return DocVec{};
+  }
+
+  static bool BySize(const IndexResult& l, const IndexResult& r) {
+    return l->size() < r->size();
+  }
+
+ private:
+  variant<DocVec /*owned*/, const DocVec* /*pointer to borrowed*/> value;
+};
+
+// Unify (OR) matched and current
+IndexResult MergeOR(IndexResult&& matched, IndexResult&& current, vector<DocId>* tmp) {
+  tmp->clear();
+  tmp->reserve(matched->size() + current->size());
+
+  set_union(matched->begin(), matched->end(), current->begin(), current->end(),
+            back_inserter(*tmp));
 
-  merge(matched.begin(), matched.end(), tmp->begin(), tmp->end(), back_inserter(*current));
-  current->erase(unique(current->begin(), current->end()), current->end());
+  IndexResult result{move(*tmp)};
+  *tmp = current.TakeOwnedOrEmpty();  // If current has a backing array, keep it
+  return result;
 }
 
-// Merge results from matched with current, keep sorted.
-void MergeResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
-  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
-  current->clear();
-  current->reserve(matched.size() + tmp->size());
+// Merge (AND) matched and current
+IndexResult MergeAND(IndexResult&& matched, IndexResult&& current, vector<DocId>* tmp) {
```

Comment:
You don't seem to use the fact that these are rvalue references (non-templated `&&`). In fact, the only one you `move()` is `tmp`, which is not one :laughing:

Reply:
(also above)